### PR TITLE
Add container based scaling to HPA

### LINKED ISSUE #86349: HPA should consider individual containers when scaling based on resource metrics
**What would you like to be added**:
Currently when calculating the utilization of resources on pods the usage of all the containers in the pods is summed and then divided by the sum of the requested resources of all containers. The HPA should have a way of specifying how the utilization should be computed. If it should be the sum of the containers, or if it should be the average of the individual container utilization or the maximum of the utilization of the containers. 

**Why is this needed**:
Consider the case when there are multiple containers in the pods, with one main container and a few sidecar containers which are over provisioned. If the main container consumes a lot of resources and is close to its requests the HPA would consider the utilization of the pod to be within limits. This would prevent scaling out. The inverse could also be true. Sidecars are sometimes necessary for the main container to function correctly. If one of the sidecars slowly down with high utilization the HPA would not scale out the target because the average utilization of the pod is still low.

### PR DESCRIPTION
**Pull request by arjunrn**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
This PR introduces a new metric source called `ContainerResourceMetricSource` and a corresponding `ContainerResourceMetricStatus` types which are used to specify metric targets for individual containers in the target pods and the status of those metrics.

**Which issue(s) this PR fixes**:

Fixes #86349

**Special notes for your reviewer**:

/label api-review

**Does this PR introduce a user-facing change?**:

```release-note
Introduces a metric source for HPAs which allows scaling based on container resource usage.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/blob/master/keps/sig-autoscaling/0001-container-resource-autoscaling.md:
- [Usage]: https://github.com/kubernetes/website/pull/23523
```

### CODE REVIEW

**Comment by liggitt** [comment]
Thanks for the PR. It looks like the linked proposal is not yet marked as implementable and is missing test plan info.

**Comment by arjunrn** [comment]
@liggitt I've updated the KEP and added the test plan as well.

**Comment by arjunrn** [comment]
@liggitt Any updates on the API review. Is there anything else I should do?

**Comment by liggitt** [comment]
Sorry, was OOO last week, have been catching up. Let's schedule a couple hours to go over the API changes... synchronous reviews tend to be quicker. I'll reach out to find a good time.

File: pkg/apis/autoscaling/types.go
Comment (liggitt) [inline pkg/apis/autoscaling/types.go:320]: this adds the possibility of several edge cases:
1. the named container do not exist in some (or all) of the selected pods
1. the named container is defined differently in some of the selected pods (as an initContainer, or without requests, or with requests)

need to document and test behavior for these scenarios

also need to explain how someone is expected to modify HPA when adding/removing/modifying a container within a deployment... HPA first, then deployment? deployment first, then HPA?

File: pkg/apis/autoscaling/types.go
Comment (liggitt) [inline pkg/apis/autoscaling/types.go:257]: if we create an HPA instance using this field in 1.19 with this PR merged, it gets serialized to a JSON annotation and persisted in v1 in etcd.

If we update this instance via a 1.18 API server via v2beta2, that API server won't know about this field, and will fail validation.

See https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#adding-unstable-features-to-stable-versions for details on how to add a new field and validation, but feature gate the new field so that we don't allow persisting data that would break an n-1 API server.

For this, we could add the field and validation in 1.19, leave the feature gate off in 1.19, then allow persisting this data in 1.20 by enabling the feature gate.

This means you'll need to enable the feature as part of tests that require persisting the object (integration tests, etc)... see use of SetFeatureGateDuringTest in the codebase for how to do that.

File: pkg/apis/autoscaling/types.go
Comment (liggitt) [inline pkg/apis/autoscaling/types.go:487]: the same feature-gate would protect this field

File: staging/src/k8s.io/api/autoscaling/v2beta1/types.go
Comment (liggitt) [inline staging/src/k8s.io/api/autoscaling/v2beta1/types.go:223]: is there a reason not to use the same schema as v2beta2 ContainerResourceMetricSource here? that would simplify (and maybe eliminate the need for) custom conversion methods. same question applies to v1 ContainerResourceMetricSource and the v1/v2beta1 ContainerResourceMetricStatus types

Comment (liggitt) [reply staging/src/k8s.io/api/autoscaling/v2beta1/types.go:223]: It looks like the schema in v2beta2 uses MetricTarget which only exists in v2beta2. Would it be worth copying that to v1/v2beta1 for use here? I don't have a strong opinion whether it's better to keep the v1/v2beta1 bit consistent with existing v1/v2beta1 metric specs, or to keep this new field consistent across versions.

File: pkg/apis/autoscaling/v2beta2/zz_generated.conversion.go
Comment (liggitt) [inline pkg/apis/autoscaling/v2beta2/zz_generated.conversion.go:518]: not sure why this changed... see if this goes back to a pointer cast after removing the `PodResource *PodResourceMetricSource`

File: pkg/apis/autoscaling/types.go
Comment (liggitt) [inline pkg/apis/autoscaling/types.go:257]: drop json tags on internal API types... they're not serializable. applies to everything in this file

File: pkg/apis/autoscaling/validation/validation.go
Comment (liggitt) [inline pkg/apis/autoscaling/validation/validation.go:402]:
```suggestion
	} else {
	 ...validate that this is a valid container name (ValidateDNS1123Label is used to validate container names in core/validation/validation.go)
	}
```

File: pkg/controller/podautoscaler/horizontal.go
Comment (liggitt) [inline pkg/controller/podautoscaler/horizontal.go:337]: defer to @kubernetes/sig-autoscaling-pr-reviews on controller review

File: pkg/controller/podautoscaler/metrics/interfaces.go
Comment (liggitt) [inline pkg/controller/podautoscaler/metrics/interfaces.go:42]: document the meaning of `container==""` in the interface godoc here

File: pkg/controller/podautoscaler/metrics/legacy_metrics_client.go
Comment (liggitt) [inline pkg/controller/podautoscaler/metrics/legacy_metrics_client.go:66]: @kubernetes/sig-autoscaling-pr-reviews

File: pkg/printers/internalversion/printers.go
Comment (liggitt) [inline pkg/printers/internalversion/printers.go:1987]: cc @kubernetes/sig-cli-pr-reviews

File: staging/src/k8s.io/kubectl/pkg/describe/describe.go
Comment (liggitt) [inline staging/src/k8s.io/kubectl/pkg/describe/describe.go:3449]: cc @kubernetes/sig-cli-pr-reviews

**Review by liggitt — COMMENTED** [review]
I think most users of the existing HPA API would be surprised that an HPA requesting 70% utilization would not scale up if one of their containers was persistently at 100% utilization. If someone's goal is to scale up if any of their containers exceed 70% average utilization, requiring switching to specify per-container metric sources is fiddly (requires putting container names in HPA, requires copy/pasting the metric spec target data, etc). In addition to per-container control, should we improve the existing resource metric source to be able to indicate it should be applied per container, rather than as a pod-level aggregate of requests/usage?

File: pkg/apis/autoscaling/types.go
Comment (arjunrn) [reply pkg/apis/autoscaling/types.go:320]: I looked into this and the controller handles this case when metrics are missing for a pod. It then fills pod's metrics cautiously. What that means is that it recomputes the recommendation with the missing pod metrics as 0 if the calculation indicates that the deployment should scale up, and as 100% utilization if the previous calculation shows that the target should be scaled down. This was added to handle the case where metrics might be missing due to broken metrics reporting on certain nodes. I'm not sure if the same behavior should be used when only some pods have the container name. 

Do you think it is sufficient if we just add a disclaimer that all the targetted pods should have the container with that name.

Also for deployment rollouts, I think it is sufficient if both the old and new container names are part of the HPA spec and after the rollout only the new container name is kept. This should be prevent unexpected scale downs.

Let me know what you think @liggitt 

cc @josephburnett

File: pkg/controller/podautoscaler/metrics/rest_metrics_client.go
Comment (arjunrn) [inline pkg/controller/podautoscaler/metrics/rest_metrics_client.go:86]: Fix the log line for when both cases: pod and container metrics are requested. Currently only handles container case.

File: pkg/apis/autoscaling/types.go
Comment (josephburnett) [reply pkg/apis/autoscaling/types.go]: Currently HPA treats missing metrics and missing request slightly differently.

If any container of a pod is missing metrics, the HPA will treat the whole pod as missing metrics: https://github.com/kubernetes/kubernetes/blob/62b8faed4042e9f1a9c1e895ad1c36bb473eb6ee/pkg/controller/podautoscaler/metrics/rest_metrics_client.go#L80-L89  In this case, as @arjunrn pointed out, we will make conservative assumptions about the metrics and proceed with autoscaling.

However if any container is missing a resource request for a configured target, HPA will refuse to take any action.  This is considered an invalid configuration (no way to calculate utilization without a denominator) so the HPA fails without taking any action.

If the user provides a container target for a container that doesn't exist, it would be more natural to treat this as an invalid configuration and do nothing.  Once the container shows up, reconciliation will proceed.  E.g. to add a new container, one should add it to the deployment and wait for all replica sets to be updated.  Then update the HPA.  In this way, there will be no period during which autoscaling is off.  However, if the user adds the container to HPA first, then the deployment, autoscaling will be suspended until all pods have the new container.  Which is not ideal, but it will stabilize to a working configuration once the deployment is finished rolling out.

File: pkg/controller/podautoscaler/replica_calculator.go
Comment (josephburnett) [inline pkg/controller/podautoscaler/replica_calculator.go:425]: As mentioned above (https://github.com/kubernetes/kubernetes/pull/90691#discussion_r463440392):

If a container is missing entirely, this will return a resource request of zero.  This is not ideal.  If the targeted container is missing, we should treat it as an invalid configuration and return an error.

**Review by josephburnett — CHANGES_REQUESTED** [review]

File: pkg/controller/podautoscaler/replica_calculator.go
Comment (arjunrn) [reply pkg/controller/podautoscaler/replica_calculator.go:425]: @josephburnett I've updated the logic to return an error in case the pod metrics endpoint does not return metrics about the the container.

**Review by josephburnett — APPROVED** [review]
These changes look good for sig-autoscaling.

However, there are still quite a few unresolved comments from @liggitt.  I think you've already started addressing those because I see code for checking the Alpha feature status.  But it would be good to go through them all and explain if / how they were resolved.

File: pkg/apis/autoscaling/types.go
Comment (liggitt) [reply pkg/apis/autoscaling/types.go:458]: in autoscalerStatusStrategy#PrepareForUpdate, follow the pattern of clearing the field if the feature is off AND the existing object does not use the new field

File: staging/src/k8s.io/api/autoscaling/v2beta1/types.go
Comment (liggitt) [reply staging/src/k8s.io/api/autoscaling/v2beta1/types.go:201]: v2beta2 metric target was expanded and unified averageUtilization/averageValue... that doesn't exist in v1 or v2beta1

File: pkg/apis/autoscaling/validation/validation.go
Comment (liggitt) [inline pkg/apis/autoscaling/validation/validation.go:366]: think about what bounds we place on all new fields. e.g. should this be length-constrained? limited to specific characters?

File: pkg/features/kube_features.go
Comment (liggitt) [inline pkg/features/kube_features.go:649]: nit: add sig-autoscaling

File: pkg/apis/autoscaling/validation/validation.go
Comment (liggitt) [inline pkg/apis/autoscaling/validation/validation.go:298]: can handle the invalid type and missing required field cases together here and make this more straightforward, in addition to providing a nicer message when the container resource feature is enabled:

```go
switch spec.Type {

case autoscaling.ObjectMetricSourceType:
  if spec.Object == nil {
    allErrs = append(allErrs, field.Required(fldPath.Child("object"), "must populate information for the given metric source"))
  }

case autoscaling.PodsMetricSourceType:
  if spec.Pods == nil {
    allErrs = append(allErrs, field.Required(fldPath.Child("pods"), "must populate information for the given metric source"))
  }

case autoscaling.ResourceMetricSourceType:
  if spec.Resource == nil {
    allErrs = append(allErrs, field.Required(fldPath.Child("resource"), "must populate information for the given metric source"))
  }

case autoscaling.ExternalMetricSourceType:
  if spec.External == nil {
    allErrs = append(allErrs, field.Required(fldPath.Child("external"), "must populate information for the given metric source"))
  }

case autoscaling.ContainerResourceMetricSourceType:
  if spec.ContainerResource == nil {
    if utilfeature.DefaultFeatureGate.Enabled(features.HPAContainerMetrics) {
      allErrs = append(allErrs, field.Required(fldPath.Child("containerResource"), "must populate information for the given metric source"))
    } else {
      allErrs = append(allErrs, field.Required(fldPath.Child("containerResource"), "must populate information for the given metric source (only allowed when HPAContainerMetrics feature is enabled)"))
    }
  }

default:
  allErrs = append(allErrs, field.NotSupported(fldPath.Child("type"), *hook.MatchPolicy, supportedMatchPolicies.List())
}
```

File: pkg/registry/autoscaling/horizontalpodautoscaler/strategy.go
Comment (arjunrn) [reply pkg/registry/autoscaling/horizontalpodautoscaler/strategy.go:87]: I'm assuming that the controller will have to handle the case where a metric source types object is present but the metric source definition is nil. Currently the controller does not do so.

Comment (arjunrn) [reply pkg/registry/autoscaling/horizontalpodautoscaler/strategy.go:87]: So it looks like after nulling out the container source it will be caught in the validate step which would throw an error. Please ignore my last comment.

File: pkg/apis/autoscaling/validation/validation.go
Comment (arjunrn) [reply pkg/apis/autoscaling/validation/validation.go:366]: I have changed it to use an existing helper method which validates that the resource is a valid container resource.

File: staging/src/k8s.io/api/autoscaling/v1/types.go
Comment (arjunrn) [reply staging/src/k8s.io/api/autoscaling/v1/types.go:209]: Added to the docs

File: pkg/features/kube_features.go
Comment (arjunrn) [reply pkg/features/kube_features.go]: Added @mwielgus and @josephburnett who are part of head and HPA in-charge respectively.

**Comment by pwittrock** [comment]
@arjunrn please split out the cli changes into a separate PR, or at least commit.

**Review by soltysh — APPROVED** [review]
sig-cli related changes lgtm

**Review by liggitt — COMMENTED** [review]
one comment on API docs, one question about the containerResource name validation, rebase, then lgtm

File: pkg/apis/autoscaling/validation/validation.go
Comment (arjunrn) [reply pkg/apis/autoscaling/validation/validation.go:390]: Right now the resource field for the regular resource type does not have any validation in it. So users can specify even `ephemeral-storage` and `hugepage-` resources for scaling. However `metrics-server` which is used by the HPA controller does not support those resources at the moment. But if they were, the HPA controller would support them out of the box because the resource name is not validated. So should I now restrict which resources can be specified(limit to cpu and memory) or should I be a bit more permissive and allow all container resources.

File: pkg/apis/autoscaling/types.go
Comment (arjunrn) [reply pkg/apis/autoscaling/types.go:458]: Updating the status is not allowed in the strategy. So for cases when the feature gate if off it will simply get copied over from the old status if it exists. And it can only exist if the controller puts it there because the feature was enabled and used.

File: pkg/apis/autoscaling/validation/validation.go
Comment (arjunrn) [reply pkg/apis/autoscaling/validation/validation.go:390]: cc @josephburnett

**Comment by liggitt** [comment]
API doc updates lgtm. I'll defer to @josephburnett on the desired allowed values for the container resource. Once that is determined, add validation unit tests demonstrating the allowed/disallowed values.

File: pkg/apis/autoscaling/validation/validation.go
Comment (josephburnett) [reply pkg/apis/autoscaling/validation/validation.go:390]: I think [replica_calculator.go](https://github.com/kubernetes/kubernetes/blob/e39bdd4ac123217a101ae5099938a566e00059b0/pkg/controller/podautoscaler/metrics/rest_metrics_client.go#L82) is pretty generic all the way through.  So we shouldn't add this validation.

Comment (arjunrn) [reply pkg/apis/autoscaling/validation/validation.go:390]: @josephburnett @liggitt In that case the current validation is correct. So I will leave this as is.

Comment (liggitt) [reply pkg/apis/autoscaling/validation/validation.go:390]: > In that case the current validation is correct

are you sure?  this validation is narrower than `validateContainerResourceName`, so it would disallow resources a container could request, right?

Comment (arjunrn) [reply pkg/apis/autoscaling/validation/validation.go:390]: I misunderstood your first comment. You're right this is more restrictive. So now I have 3 options:
1. Copy the `validateContainerResourceName` function. 
2. Turn the `validateContainerResourceName` into a public function and add tests.
3. Remove the validation to keep it the same as validation for the resource name in the `ResourceMetricSource`. 

I prefer option 2. WDYT?

Comment (liggitt) [reply pkg/apis/autoscaling/validation/validation.go:390]: 2 sounds good to me

Comment (arjunrn) [reply pkg/apis/autoscaling/validation/validation.go]: @liggitt I've made changes. Can you take another look.

**Comment by liggitt** [comment]
thanks, changes look good

**Pull request merged**
